sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala

```diff
@@ -408,7 +408,7 @@ trait HiveTypeCoercion {
         Union(newLeft, newRight)
 
       // fix decimal precision for expressions
-      case q => q.transformExpressions {
+      case q => q.transformExpressionsUp {
```

Comment:
Since we separate type checking logic from `resolved`,  we need to ensure all children have valid input types before we do type coercion(think about `Add(Add(decimal1, decimal2), decimal3)`). Here I just use `transformExpressionsUp` as a quick fix(and only for `DecimalPrecision` rule as it breaks a test...), should we skip not only `!childrenResolved` but also `!children.forall(_.validInputTypes)` during all type coercion, or still checking types in `resolved`? @rxin @marmbrus